rasa_sdk/knowledge_base/actions.py

```diff
@@ -125,20 +125,35 @@ async def run(
         object_type = tracker.get_slot(SLOT_OBJECT_TYPE)
         last_object_type = tracker.get_slot(SLOT_LAST_OBJECT_TYPE)
         attribute = tracker.get_slot(SLOT_ATTRIBUTE)
+        has_mention = tracker.get_slot(SLOT_MENTION) is not None
 
         new_request = object_type != last_object_type
 
         if not object_type:
-            # object type always needs to be set as this is needed to query the
-            # knowledge base
-            dispatcher.utter_message(response="utter_ask_rephrase")
-            return []
+            # sets the object type dynamically from entities if object_type is not found
+            # in user query
+            object_types = await utils.call_potential_coroutine(
+                self.knowledge_base.get_object_types()
+            )
+            object_type_dynamic = await get_object_type_dynamic(tracker, object_types)
 
-        if not attribute or new_request:
-            return await self._query_objects(dispatcher, object_type, tracker)
-        elif attribute:
+        if object_type:
+            if not attribute or new_request:
+                return await self._query_objects(dispatcher, object_type, tracker)
+
+            elif attribute:
+                return await self._query_attribute(
+                    dispatcher, object_type, attribute, tracker
+                )
+
+        if object_type_dynamic and attribute:
             return await self._query_attribute(
-                dispatcher, object_type, attribute, tracker
+                dispatcher, object_type_dynamic, attribute, tracker
+            )
+
+        if last_object_type and has_mention and attribute:
+            return await self._query_attribute(
+                dispatcher, last_object_type, attribute, tracker
```

Comment:
In which scenarios is this new code required?

Reply:
In rasa’s current implementation, the object_type slot is always set. However, in our solution, it is sometimes determined during execution. This requires the object_type slot to be a None value. However, when a user inquires about an object using `mention`, for example, ‘what is the cuisine of the second one?’, we must pass in the `last_object_type` value to `self._query_attribute()`. This was not required in the original Rasa implementation, as the object_type was directly taken from the slot.

Reply:
Ok got it, but how do we know that the mention refers to the last_object_type (not a trick question!) 🤔 
Also please add a test for this new if branch.

Reply:
The 3rd test case in `test_actions.py` checks for this new if branch when action is run.

Regarding the clarification on `mention` and `last_object_type`, in Rasa’s existing implementation,  two types of `mention` scenarios are considered:

1. Ordinal mention: When a user asks, “Can you tell me the price range of the first one?”, it is assumed that the user is asking about the first object from the recent list of objects. For example, when a user asks “list me some restaurants”, followed by “Can you tell me the price range of the first one?” ,the knowledge base action will query for the `price range` of the first object displayed in the list of restaurants. For this purpose, it is assumed that the user is referring to the `object` in the `last_object_type` slot in that conversation. 
2. Other mention: When a user asks for an object attribute using mention words such as "it" or "that one", it is assumed that the user is referring to the last mentioned `object` in that conversation which is stored in `last_object` slot. In that case, the function checks if `last_object_type` slot is not None. If so, the function extracts the attribute of the object in the `last_object` slot.

In both situations, the implementation assumes that the user is talking about an object belonging to the `last_object_type`. This is the default assumption in rasa knowledge base actions.

Reply:
All clear now; to double-check, are you referring to test case index `2` or test case index `3` in the list of parameter test cases?

Reply:
test case index `3`. I have copied it here for your immediate reference.

```
(
    {
        "entities": [
            {
                "entity": "attribute",
            },
            {
                "entity": " mention",
            },
        ],
    },
    {
        SLOT_MENTION: "2",
        SLOT_ATTRIBUTE: "cuisine",
        SLOT_OBJECT_TYPE: None,
        SLOT_LISTED_OBJECTS: [1, 2, 3],
        SLOT_LAST_OBJECT: None,
        SLOT_LAST_OBJECT_TYPE: "restaurant",
    },
    [
        SlotSet(SLOT_MENTION, None),
        SlotSet(SLOT_ATTRIBUTE, None),
        SlotSet(SLOT_OBJECT_TYPE, None),
        SlotSet(SLOT_LAST_OBJECT, 2),
        SlotSet(SLOT_LAST_OBJECT_TYPE, "restaurant"),
    ],
)
```